subsys/bluetooth/services/ots/ots.c

```diff
@@ -495,6 +546,65 @@ BT_GATT_SERVICE_INSTANCE_DEFINE(ots_service_list, ots_instances,
 				CONFIG_BT_OTS_MAX_INST_CNT,
 				BT_GATT_OTS_ATTRS);
 
+static void ots_delete_empty_name_objects(struct bt_ots *ots)
+{
+	char id_str[BT_OTS_OBJ_ID_STR_LEN];
+	struct bt_gatt_ots_object *obj;
+	struct bt_gatt_ots_object *next_obj;
+	int err;
+
+	err = bt_gatt_ots_obj_manager_first_obj_get(ots->obj_manager, &next_obj);
+	while (!err) {
+		obj = next_obj;
+
+		/* Get the next object before we potentially delete the current object and
+		 * no longer can get the next object
+		 */
+		err = bt_gatt_ots_obj_manager_next_obj_get(ots->obj_manager, obj, &next_obj);
+
+		if (strlen(obj->metadata.name) == 0) {
+			bt_ots_obj_id_to_str(obj->id, id_str, sizeof(id_str));
+			LOG_DBG("Deleting object with %s ID due to empty name", log_strdup(id_str));
+
+			if (ots->cb->obj_deleted) {
+				ots->cb->obj_deleted(ots, NULL, obj->id);
+			}
+
+			if (IS_ENABLED(CONFIG_BT_OTS_DIR_LIST_OBJ)) {
+				bt_ots_dir_list_obj_remove(ots->dir_list, ots->obj_manager,
+							   ots->cur_obj, obj);
+			}
+			bt_gatt_ots_obj_manager_obj_delete(obj);
+		}
+	}
+}
+
+static void ots_conn_disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	uint32_t index;
+	struct bt_ots *instance;
+
+	for (instance = BT_GATT_OTS_INSTANCE_LIST_START, index = 0;
+	     index < instance_cnt;
+	     instance++, index++) {
+
+		LOG_DBG("Processing disconnect for OTS instance %u", index);
+
+		if (instance->cur_obj != NULL) {
+			__ASSERT(instance->cur_obj->state.type == BT_GATT_OTS_OBJECT_IDLE_STATE,
+				"The current object is expected to be in idle state as part "
+				"of cleanup of the L2CAP channel connection close.");
+			instance->cur_obj = NULL;
+		}
+
+		ots_delete_empty_name_objects(instance);
```

Comment:
This honestly seem like a potential very expensive operation. 

On any disconnect we go through all the objects of all the OTS instances to find those with non-initialized names. It probably works, but if the server has many objects (hundreds or thousands of objects) for each instance, this will be quite expensive, and has a complexity of `O(M * N)` where `M` is number of instances and `N` is the number of objects, and is run for each disconnect. 

It is fine for now I suppose, but we could consider keeping a small cache of of the objects that is missing a name instead, and only loop through those, which will likely be much much smaller.

Reply:
Yeah, this is not a very efficient operation as you have pointed out. I did consider the cache but we have a similar operation during the object name write procedure where we have to check all other object names for duplication so I left this similar to that.

Reply:
That's a good point. 

If we implement some way of quickly looking up string duplicates for OTS, then we can reuse that to check for the empty string as well. An obvious solution for that is of course a hashmap of the names, but as with everything else, there's a tradeoff between computation and memory :) 

Alternatively I guess our server could impose an additional requirement that no more than 1 object may be without name, and then just have a single reference to that object until it is given a unique name.

Reply:
I think for now I will leave this computationally intensive solution in place. We can optimize this later if we find it a problem.